lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -445,61 +406,139 @@ function Get-CharmServices {
             )
         }
     }
-
+    $s2dContextGen = @{
+        "generator" = (Get-Item "function:Get-S2DCSVContext").ScriptBlock
+        "relation" = "csv"
+        "mandatory" = $false
+    }
+    if((Confirm-JujuRelationCreated -Relation 'csv')) {
+        $s2dContextGen['mandatory'] = $true
+    }
+    $freeRdpContextGen = @{
+        "generator" = (Get-Item "function:Get-FreeRDPContext").ScriptBlock
+        "relation" = "free-rdp"
+        "mandatory" = $false
+    }
+    if((Confirm-JujuRelationCreated -Relation 'free-rdp')) {
+        $freeRdpContextGen['mandatory'] = $true
+    }
+    $jujuCharmServices['nova']['context_generators'] += @($s2dContextGen, $freeRdpContextGen)
     return $jujuCharmServices
 }
 
 function Get-FreeRDPContext {
     Write-JujuWarning "Getting context from FreeRDP"
-
     $required = @{
         "enabled" = $null
         "html5_proxy_base_url" = $null
     }
-
     $ctx = Get-JujuRelationContext -Relation "free-rdp" -RequiredContext $required
     if (!$ctx.Count) {
         return @{}
     }
-
     return $ctx
 }
 
-function Get-S2DContext {
-    Write-JujuWarning "Generating context for S2D"
+function Get-S2DVolumeName {
+    $volumeName = (Get-JujuLocalUnit) -replace '/', '-'
+    return $volumeName
+}
 
-    $version = Get-OpenstackVersion
-    if (!$NOVA_PRODUCT[$version]['compute_cluster_driver']) {
-        Write-JujuWarning "Hyper-V Cluster driver is not supported for release '$version'"
-        return @{}
-    }
+function Confirm-RunningClusterService {
+    Start-Service "ClusSvc"
```

Comment:
I do the check for both relations `csv` and `s2d` [here](https://github.com/ionutbalutoiu/nova-hyperv-charm/blob/master/lib/Modules/ComputeHooks/ComputeHooks.psm1#L706). If those relations are fine, it means I have a cluster. This function is not exported, so I consider it as a `private` function, thus I don't need to double check that here as well.